pkg/ebpftracer/c/tracee.bpf.c

```diff
@@ -5442,141 +5443,85 @@ statfunc u32 submit_netflow_event(struct __sk_buff *ctx, net_event_context_t *ne
-                if (!netflowvalptr) {
-                    return 0;
-                }
-            }
-
-            update_flow_stats(ctx, netflowvalptr, retval_hasflag(packet_ingress));
-
+        {
             u64 now = bpf_ktime_get_ns();
-            u64 last_submit_seconds = (now - netflowvalptr->last_update) / 1000000000;
+            u64 last_submit_seconds = (now - value->last_update) / 1000000000;
             // Check if it's time to submit flow sample.
             if (last_submit_seconds >= global_config.flow_sample_submit_interval_seconds) {
-                netflowvalptr->last_update = now;
-                submit_netflow_event(ctx, neteventctx, netflowvalptr);
-                reset_flow_stats(netflowvalptr);
-                return 0;
+                value->last_update = now;
+                submit_netflow_event(skb, neteventctx, value);
+                reset_flow_stats(value);
             }
-
-            // Flow sample should not be submitted yet, exit.
             return 0;
-
-        // 3) TCP connection is being closed/terminated.
+        }
         case flow_tcp_end:
-            // Any side can close the connection (FIN, RST, etc). Need heuristics.
-
-            // Attempt 01: Try to find the flow using current src/dst.
-            for (int n = 0; n < 3; n++) {
-                netflowvalptr = bpf_map_lookup_elem(&netflowmap, &neteventctx->md.flow);
-                if (!netflowvalptr)
-                    continue;
-            }
-
-            // FIN could be sent by either side, by both, or by none (RST). Need heuristics.
-            if (!netflowvalptr) {
-                // Attempt 02: Maybe this packet src wasn't the flow initiator, invert src/dst.
+        {
+            u64 active = __sync_fetch_and_sub(&value->active, 1) -1;
+            if (active <= 0) {
+                submit_netflow_event(skb, neteventctx, value);
+                bpf_map_delete_elem(&netflowmap, &neteventctx->md.flow);
                 neteventctx->md.flow = invert_netflow(neteventctx->md.flow);
-
-                for (int n = 0; n < 3; n++) {
-                    netflowvalptr = bpf_map_lookup_elem(&netflowmap, &neteventctx->md.flow);
-                    if (!netflowvalptr)
-                        continue;
-                }
-
-                // After first FIN packet is processed the flow is deleted, so the second
-                // FIN packet, if ever processed, will not find the flow in the map, and
-                // that is ok.
-                if (!netflowvalptr)
-                    return 0;
-
-                // Flow was found using inverted src/dst: current pkt dst was the flow initiator.
-                is_initiator = 0;
-
-            } else {
-                // Flow was found using current src/dst: current pkt src was the flow initiator.
-                is_initiator = 1;
+                bpf_map_delete_elem(&netflowmap, &neteventctx->md.flow);
```

Comment:
Mhm we only ever initialize the value in the `netflowmap` with the value we find in this method. I guess there is no reason for inverting the flow and trying to delete both variants here.